File: Recursion/Leetcode/Medium/Subsets/Subsets.cpp
// https://leetcode.com/problems/subsets/description/
#include<bits/stdc++.h>
using namespace std;
void subSequence(vector<int> &nums, vector<int> &sub, vector<vector<int>> &allSubs, int index)
{
    if (index >= nums.size())
    {
        allSubs.push_back(sub);
        return;
    }
    subSequence(nums, sub, allSubs, index + 1);
    sub.push_back(nums[index]);
    subSequence(nums, sub, allSubs, index + 1);
    sub.pop_back();
}
vector<vector<int>> subsets(vector<int> &nums)
{
    vector<vector<int>> allSubs;
    vector<int> sub;
    subSequence(nums, sub, allSubs, 0);
    return allSubs;
}
int main(void) {
    vector<int> num = {1, 2, 3};
    vector<vector<int>> vec = subsets(num);
    for(auto v : vec) {
        for(int ele : v) {
            cout << ele << " ";
        }
        cout << endl;
    }
    return 0;
}

File: Recursion/Leetcode/Medium/Subsets/Subsets.java
// https://leetcode.com/problems/subsets/description/
import java.util.ArrayList;
import java.util.List;

public class Subsets {
    static void generate(int[] nums, int index, List<Integer> seq, List<List<Integer>> combo) {
        if (index == nums.length) {
            combo.add(new ArrayList<>(seq));
            return;
        }
        generate(nums, index + 1, seq, combo);
        seq.add(nums[index]);
        generate(nums, index + 1, seq, combo);
        seq.remove((int) (seq.size() - 1));
    }

    static List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> seq = new ArrayList<>();
        generate(nums, 0, seq, ans);
        return ans;
    }

    public static void main(String[] args) {
        List<List<Integer>> lists = subsets(new int[]{1,2,3});
        System.out.println(lists);
    }
}
